sycl/include/sycl/ext/intel/esimd/memory.hpp

```diff
@@ -529,17 +531,23 @@ gather_impl(AccessorTy acc, simd<uint32_t, N> offsets, uint32_t glob_offset,
 ///   predicate are not accessed, their values in the resulting vector are
 ///   undefined.
 ///
-template <typename T, int N, typename AccessorTy>
-__ESIMD_API std::enable_if_t<(sizeof(T) <= 4) &&
-                                 (N == 1 || N == 8 || N == 16 || N == 32) &&
-                                 !std::is_pointer<AccessorTy>::value,
-                             simd<T, N>>
-gather(AccessorTy acc, simd<uint32_t, N> offsets, uint32_t glob_offset = 0,
+template <typename T, int N, typename AccessorTy, typename Toffset,
+          typename Tgloboffset = uint32_t>
+__ESIMD_API std::enable_if_t<
+    (sizeof(T) <= 4) && (N == 1 || N == 8 || N == 16 || N == 32) &&
+        !std::is_pointer<AccessorTy>::value && std::is_integral_v<Toffset> &&
+        std::is_integral_v<Tgloboffset>,
+    simd<T, N>>
+gather(AccessorTy acc, simd<Toffset, N> offsets, Tgloboffset glob_offset = 0,
        simd_mask<N> mask = 1) {
 #ifdef __ESIMD_FORCE_STATELESS_MEM
   return gather<T, N>(__ESIMD_DNS::accessorToPointer<T>(acc, glob_offset),
                       offsets, mask);
 #else
+#ifdef __SYCL_DEVICE_ONLY__
```

Comment:
the reason we need this ifdef is because of the host compile. `__ESIMD_FORCE_STATELESS_MEM` is not set there, so we would hit this assert.

i wonder why i had to do this but never saw it done elsewhere, it seems this is the first case of trying to do something like this. if im wrong and theres a better way to solve the host compile problem please let me know